Hover info for local bindings

ChrisPenner opened the pull request:
## Overview

* Wires up Hover in LSP to provide this type info when hovering on references to local-bindings, or binding sites

<img width="276" alt="image" src="https://github.com/unisonweb/unison/assets/6439644/db07df75-e7cf-4f9d-8a77-7cd0227b3638">

Note: Doesn't currently work for top-level binding locations or Pattern binding locations.

## Implementation notes

* Rather than only emitting notes for top-level bindings in the type-checker, emit notes for ALL bindings by detecting "Ann" nodes going out of scope in the typechecker context
* Utilize the tooling in place for typechecking blanks to also keep track of solved type info for the local bindings.
* Wire up hover so if the hover is a var or a binding location, check the local binding types and find a match.

## Test coverage

* [x] Tests for hover info.

## Loose ends

* [x] Support hovering the name of a binding: 
  * The LHS of bindings are converted into `Abs` nodes in the resulting tree, and then are moved around all over the place in the tree as part of normalization, This makes it tricky to determine which variable we're referencing when hovering over the var at the binding site; ironically we can find the type of a var, just not at it's definition point 😆 . I can probably fix this either by keeping more info in the tree, or adding some special cases to how the LSP crawls the ABT, but it'll take a little more time to sort out.
* [x] Support hovering variables bound as function arguments
  * I should be able to emit notes about these during typechecking too. Just need to track down the right spot for it.
* [x] Ensure we handle letrec's correctly, I suspect it doesn't work in some cases.

To come later:
* The information we now track enables 'jump-to-definition' for local bindings
* Add type on hover for pattern binding locations
* Add type on hover for top-level bindings

ChrisPenner committed:
Initial local var notes working

ChrisPenner committed:
First pass at local var binding type

ChrisPenner committed:
Cleanup

ceedubs commented:
Hmm I feel like I'm doing something wrong here, but when I try this out I get a totally different type than in your screenshot.

![image](https://github.com/unisonweb/unison/assets/977929/d1bbec73-df8c-46d2-b7ab-03c32c6b490c)

Also I just get "No information found" for local variables in the more involved definition that I tried. Any idea why our results might differ? I'm running a local build of 00d32cc4f .

ChrisPenner commented:
@aryairani  P.s. I don't think this is ready for review from anyone yet btw; I still have to fix let-recs, figure out what's up with Cody's issue, and fix some annotations 😄

ChrisPenner commented:
@ceedubs  Hrmm, I get your result if I remove the `Nat.` from the `a Nat.+ 2`;
Looks like I might be grabbing types from before TNDR has done its thing, probably fixable 😄

ceedubs commented:
Oh I didn't realize that I failed to copy verbatim. Something something jpegs are lossy when copy/pasting code. But I guess that it was good because I might have caught an issue :)

ChrisPenner commented:
No matter how much I work on something I can always trust @ceedubs to find something I missed, which I actually really appreciated 😄 ;
Something to be said for having someone really unlucky on the team 😅 😆

ceedubs commented:
@ChrisPenner I don't have a sense of how much work it is to get this to a state in which you think that it is ready for review. It's certainly not a high priority, but I just wanted to let you know that this would be a very welcome feature if you think that it's not much work to push it through :)

ChrisPenner committed:
Improve local binding inference

ChrisPenner committed:
Add annotations for lambda bindings

ChrisPenner committed:
Fix hover info for local bindings

ChrisPenner committed:
Don't typecheck twice, correctly find best match for each var.

ChrisPenner committed:
Get letrec bindings working in functions

ChrisPenner committed:
disable binding types

ChrisPenner committed:
Revive top-level var bindings

ChrisPenner committed:
Remove unused libs

ChrisPenner committed:
Re-merge trunk

ChrisPenner committed:
Do var reporting inside retract

ChrisPenner committed:
Really naive version working

Need to fix case where multiple vars with the same name exist
Need to fix argument binding detection

ChrisPenner committed:
Debugging

ChrisPenner committed:
Working relatively well IFF you have a top-level signature

ChrisPenner committed:
Note var mentions so we can align symbol ids with their location then tie that to their type.

ChrisPenner committed:
Pass along binding symbol annotations and emit notes for them to LSP layer

ceedubs commented:
@ChrisPenner I've been using this branch (with some updates from trunk) for a couple of days and it has been great!

[Here](https://share.unison-lang.org/@unison/cloud/code/releases/20.2.1/latest/terms/Cloud/run/local/impl) is an example that I noticed where the type doesn't look right. It shows `'{IO} 𝕣`, but the actual type should be `Signal [Text]`, which is annotated a few lines above. I thought that maybe it was finding a `stdout` term in base, but that doesn't seem to be the case because those are named `stdOut` and `StdOut` (and don't have this type signature either).

![image](https://github.com/user-attachments/assets/6ee4da8b-bb71-47b9-ae83-c213eaaadca7)

aryairani commented on .vscode/settings.json:
checking if checking this in was intentional bc it wasn't mentioned in the PR desc

aryairani commented on .gitignore:
I wonder if there's some syntax for these without the exceptions, like `/*.u`?

ChrisPenner commented:
@aryairani  ugh, I think the diff got all mixed up; there's definitely not this many actual changes.

pchiusano commented on parser-typechecker/src/Unison/Typechecker/Context.hs:
```diff
@@ -521,7 +530,18 @@ markThenRetract hint body =
     r <- body
     ctx <- retract
-    pure ((r, ctx), substituteSolved ctx)
+    let solvedCtx = substituteSolved ctx
+    for_ ctx \case
+      var@(Ann v typ) -> do
+        Debug.debugM Debug.Temp "Ann" var
+        noteVarBinding v  (TypeVar.lowerType typ)
+      v@(Var{}) ->
+        Debug.debugM Debug.Temp "Var" v
+      (Solved _ v t) -> do
+        Debug.debugM Debug.Temp "Solved" v
+        noteVarBinding v (TypeVar.lowerType $ Type.getPolytype t)
+      _ -> pure ()
+    pure ((r, ctx), solvedCtx)
```
This doesn't seem right to me... or it might not be wrong but could lead to quadratic performance.

I think you want to emit notes when things are solved. This seems like it will be traversing the same context over and over again.

ChrisPenner replied:
Yup, this is the sort of thing I had in mind when I'd requested a bit of time from You or Dan to look through this PR a while back, the PR will be stuck on hold until either someone has a couple minutes to sit down with me and walk through a few things, or until I can dedicate sufficient time to dig through and learn how the typechecker is set up, which might be good for knowledge sharing anyways :)

ChrisPenner committed:
Merge trunk

ChrisPenner committed:
Try what Dan recommended

ChrisPenner commented:
Notes on the current status:

* After a chat with Dan we think the required machinery exists, and it must, since we use it for solving "blanks" (a.k.a. vars which may be solvable via TDNR). There is however some weirdness with Universal and Existential types in this form that might need another look.
* Also I don't yet support getting the type of a declaration itself, only usage sites.
* Also for some reason it only seems to work with a top-level signature? which is strange...

ChrisPenner commented:
For Cody's example, I believe what's actually happening is it's showing the type of some other thing in scope and not the stdout, e.g. it's showing the type of the do-block or something.

This is definitely possible since we actually wire up types by _position_ rather than symbol or name.
I was unable to minimize that example, but will revisit later.

ChrisPenner commented:
Okay here's a minimal reproduction of the issue, hovering over `myid` gives the type of `rec` instead; it appears to only be an issue when in a self-recursive function, I think it's somehow treating every symbol in the body of the func as within the scope of the "local binding" rec, which actually isn't a local binding at all.

```
rec = do
    use Text ++
    myid x = x
    myVal = myid "" ++ rec Null
    myVal
```

ChrisPenner committed:
Remove binding locations as type mentions for now

ChrisPenner commented:
It's still not quite on things like this:

```
xyz : x -> x
xyz  bar = bar
```

<img width="230" alt="image" src="https://github.com/user-attachments/assets/00891ce6-c831-444c-8205-bf8a89d5ad15" />


I wonder if maybe we're not doing an "applyCtx"  / solveBlanks after retracting on top-level scopes?

ceedubs reviewed (approved):
It sounds like there are some cases where it doesn't show quite as helpful of info as one would like, but after using this for an hour or two with the most recent changes it seems helpful enough to ship in its current form to me :tada: 

One thing that I noticed is that hover info doesn't show up at the definition site itself, only on subsequent references. This is also true for arguments to functions. Is it easy to expose it at the definition site too?

ChrisPenner committed:
Try to _do it right_: working for var mentions

ChrisPenner commented:
Variable mentions seem to be working well!

Now need to figure out binding locations

Ideas for getting types at binding locations:

* Inside freshen, record the mapping from v' -> v so if on hover I find a v' I can always normalize to the unfreshened symbol
* Maybe I can get the location of the LHS of bindings by taking the annotation of the top-level Abs node, and diffing it with the body location, to get just the name location

ChrisPenner committed:
Experimentation

ChrisPenner committed:
(failed) Try to fix binding annotations, even if they're not nested

ChrisPenner committed:
Clean up minimizations

ChrisPenner committed:
Attach var names to abs notes properly

ChrisPenner committed:
WIP

ChrisPenner committed:
Constant local term definition sites working

ChrisPenner committed:
Fix "ignored" actions

ChrisPenner committed:
Fix lambda bindings

ChrisPenner committed:
Try to fix "do" scoping

ChrisPenner committed:
Working for things except lambda binding sites

ChrisPenner committed:
Generalize node searching

ChrisPenner committed:
Remove debugging logs

ChrisPenner committed:
Markdownify rendering

ChrisPenner committed:
Merge trunk

ChrisPenner committed:
Clean up duplicate combinators

ChrisPenner committed:
Add tests for hover info

ChrisPenner commented:
Locally this seems to hit intermittent test failures with "clearWatchCache" tests, which I don't understand since I didn't touch anything that should affect those; I'm going to guess it's either something on trunk, or my local test build is screwed up in some way 🤔

ChrisPenner commented on parser-typechecker/src/Unison/Syntax/TermParser.hs:
```diff
@@ -1537,19 +1572,21 @@ block' isTop implicitUnitAtEnd s openBlock closeBlock = do
-                      (ann tm <> ann result)
-                      (ann tm, positionalVar (ann tm) (Var.named "_"), tm)
+                      fullLetRecSpan
+                      (Ann.External, positionalVar (ann tm) (Var.named "_"), tm)
```
We really shouldn't be using `User` vars for generated bindings like this; but at the very least I need to mark it as External to the file.

aryairani replied:
Agree about the `User` vars, but it's okay to fix it separately.

ChrisPenner commented on parser-typechecker/src/Unison/Typechecker/Components.hs:
```diff
@@ -35,25 +34,26 @@ ordered = ABT.orderedComponents
-          . sortBy
-            (compare `on` fst)
-      grouped = group bindings
-      dupes = filter ok grouped
+  let bindings :: [(v, Term' vt v a)]
+      bindings = bs <&> \((_a, v), t) -> (v, t)
+      grouped :: Map v (NESet a)
+      grouped =
+        bs
+          & fmap (\((a, v), _t) -> (v, NESet.singleton a)) -- For duplicates, we care about the binding location.
+          & Map.fromListWith (NESet.union)
+      dupes = Map.filterWithKey ok grouped
         where
-          ok (v, as)
+          ok v as
             | Var.name v == "_" = False
             | otherwise = length as > 1
-   in if not $ null dupes
-        then Left $ Nel.fromList dupes
-        else
+   in case NEL.nonEmpty $ Map.toList dupes of
+        Just dupeList -> Left dupeList
+        Nothing -> do
```
At one point I thought minimization was affecting how lets were being constructed; it turns out to be unrelated but I had a tough time reading this function as it was, so made a couple changes to help readability as well as to use stronger types like NESet instead of `[a]`

ChrisPenner commented on parser-typechecker/src/Unison/Typechecker/Context.hs:
```diff
@@ -380,9 +390,11 @@ substituteSolved ::
-substituteSolved _ i = i
+substituteSolved ctx  = \case
+  (SolvedBlank b v t) -> SolvedBlank b v (applyCtx ctx t)
+  VarBinding v loc t -> VarBinding v loc (applyCtx ctx t)
```
Here's where type info is propagated to the tracked Var bindings as well

pchiusano reviewed (approved):
Nice!!

aryairani commented:
This is exciting but what do you think is up with the transcript failures?

ceedubs commented:
Those errors seem to be 503s when transcripts are calling Share. It looks like at least one transcript might be trying to pull `unison/http`?

So these are probably just because of the issues that Share was having yesterday and not because of this PR. Maybe we can re-run them and see if they work this time.

ceedubs commented:
I rebuilt and the errors that show up now seem like they probably _are_ related to this PR. I see an LSP fold ranges transcript failing.

ChrisPenner committed:
Clean up file annotations again

ChrisPenner committed:
Change which annotations we use for formatting and folding ranges

ChrisPenner commented:
Yeah, some of the Annotation changes seem to have affected error printouts or pretty-printing somehow 🤔 

Will investigate 🔍

ChrisPenner committed:
Re-run transcripts

ChrisPenner commented on unison-src/transcripts/idempotent/duplicate-names.md:
I talked over this diff with @aryairani , the new annotation ranges I'm using for var names mean we can now render _just_ the actual binding name location, not the whole term. 

I could maintain trunk behaviour here, but after talking it over we liked the concise form better, it's less noisy than showing the entire (possibly huge) term, and instead we just show the actual line where the binding is defined.

aryairani commented on parser-typechecker/src/Unison/UnisonFile/Type.hs:
```diff
@@ -17,7 +17,7 @@ import Unison.WatchKind (WatchKind)
   { dataDeclarationsId :: Map v (TypeReferenceId, DataDeclaration v a),
     effectDeclarationsId :: Map v (TypeReferenceId, EffectDeclaration v a),
-    terms :: Map v (a {- ann for whole binding -}, Term v a),
+    terms :: Map v (a {- ann for name of the binding -}, Term v a),
```
Oooh, I hope changing this here doesn't cause surprises later; e.g. if we want to slice up and reorder the scratch file, we won't be able to use this annotation to do it if it's just the name.

ChrisPenner replied:
Maintaining the ability to do that sort of thing is covered by the auto-format transcripts which do exactly what you suggest (that is, they rewrite the scratch file by replacing each function span with the formatted text of that function).

aryairani replied:
Ah great thanks

aryairani commented on unison-cli/src/Unison/Codebase/Editor/HandleInput/FormatFile.hs:
```diff
@@ -82,8 +82,9 @@ formatFile makePPEDForFile formattingWidth currentPath inputParsedFile inputType
     (FileSummary.termsBySymbol fileSummary)
-      & Map.filter (\(_bindingAnn, _, trm, _) -> shouldFormatTerm (ABT.annotation trm) trm)
-      & itraverse \sym (_bindingAnn, mayRefId, trm, _typ) -> do
+      & fmap (\(bindingAnn, mayRefId, trm, _typ) -> (bindingAnn <> ABT.annotation trm, mayRefId, trm))
```
Or, I guess you can recover it from the term later?

ChrisPenner replied:
Yeah the full span annotation can be recovered just like you see here, nothing is lost :) 

But in case you're wondering, I'm doing all sorts of very creative things to get as much annotation information as I can from a limited number of annotation slots; it's holding together so far, but ideally I'd add a lot more slots and label them better, but I can't do that inside terms without doing a huge migration :'(

aryairani replied:
We can take a look at what it would take to do a migration, I don't think we should generally super scared of them; only scared on a case by case basis ideally :)

ChrisPenner replied:
Well, we've never done a codebase migration where Share is involved, so I'm a little scared at least 🙈
